### PR TITLE
[OptionsResolver] Add support for a failsafe mode.

### PR DESCRIPTION
**Pull request by wpottier**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | 
| License       | MIT
| Doc PR        | n/a

The failsafe mode will allow some validation errors to be ignored when resolving options.
Ignored errors in failsafe is:
 - extra options (option will be skipped)
 - unallowed type if a valid default value is provided (default value will be used)
 - unallowed value if a valid default value is provided (default value will be used)

To enable failsafe, call the enableFailsafe() method.
```php
$optionResolver = new OptionResolver();
$optionResolver->enableFailsafe();
```

### CODE REVIEW

**Comment by ro0NL** [comment]
At least for `
unallowed type/value if a valid default value is provided` id say dont pass anything invalid then, try/catch it or dont set allowed values. This hacks the design of the component IMHO.

`extraValuesAllowed` might be a new option or so, but same applies really; dont pass, try/catch or define those.

**Comment by wpottier** [comment]
@ro0NL i have a concrete example of usage, in my team we manage a press website where some half advanced user configure some blocks on the website with json data. We use the OptionResolver to ensure that the provided configuration match with what the block expect (multiple block type with each one is own configuration schema). What we want with this enhancement is to have this "failsafe" behavior for our production env. Currently, if there is an error in the configuration the block does not render at all in production (generate an error on other env). With this PR we could render the block by ignoring some errors.
Another example is when we prepare the release of a new configuration schema. OptionResolver is often failing to hard despite the fact the block as enough information with is default configuration to be rendered.

The purpose is not to hack the component but to be able to configure a sort of error level we could accept with it.

**Comment by wpottier** [comment]
hope my poor english level is still understandable ;-)

**Comment by ro0NL** [comment]
Sure :) that's what english is for.

Case is clear, yet that's still all your app logic. IMO. the term "failsafe" within the component is to vague; is it a try/catch for all failures? It doesnt tell that.

Basically component exposes it's defined/required options, allowing you to go "failsafe". For allowed values/types id say; try/catch. Or keep a ref of what you passed previously.

**Comment by ro0NL** [comment]
Im looking at the code a bit more, but find it a bit complex / hard to follow.

So before i go too fast; can you make tests pass (your usecases) without the change, in userland? Maybe using this hidden gem https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Form/Util/OptionsResolverWrapper.php :)

If not; what's not possible?

**Comment by jvasseur** [comment]
@ro0NL I don't think it is possible yet in userland (without copying a lot of code).

But with a small change to the OptionResolver class (https://github.com/jvasseur/symfony/commit/bb820c2e12693c2e3cd0ea6a108196a4f861ec78) it became possible with a class like this:

```
class FailSafeOptionsResolver extends OptionsResolver
{
    private $defaults;

    public function setDefault($option, $value)
    {
        $this->defaults[$option] = $value;

        return parent::setDefault($option, $value);
    }

    public function resolve(array $values)
    {
        return parent::resolve(array_intersect_key($values, array_flip($this->getDefinedOptions())));
    }

    public function offsetGet($option)
    {
        try {
            return parent::offsetGet($option);
        } catch (\Exception $e) {
            if (array_key_exists($option, $this->defaults)) {
                return $this->defaults[$option];
            }

            throw $e;
        }
    }
}
```

It's not a 100% working solution: it doesn't handle lazy options and default values when used as fallback are not validated/normalized but for most cases it could probably be ok or be updated to work.

This could event be simpler by adding a getter for an options default value to the `OptionsResolver` class. But I'm not sure what this getter should return for lazy defaults ?

**Comment by ro0NL** [comment]
Exposing default values sounds nice :+1: silently ignoring _passed_ values as a feature not so much :) But should be possible to do with a parent-class trick.

**Comment by sstok** [comment]
Allowing invalid values and using default? 😐 someone has been doing to much MySQL :trollface:

👍 to add an option that allows to ignore unknown options, but 👎💯 for falling back to default when the value is invalid. If your users are able to provide invalid data and then crash the application this is change is nothing more then a cheap hack. And it will cause allot of WTF when someone isn't getting an error but instead see a different value (_hello MySQL date `0000-00-00 00:00:00`_) and wonder why there value is not used. Errors like a seatbelt, they protect you when needed. When you remove them there it's no longer safe to trust your data, as it's invalid but it's changed, but that's not what the user provided.

To introduce this little use-case to the Symfony framework feels wrong, and you should properly look into a different solution 😃

**Comment by fabpot** [comment]
Closing for the reasons given by @sstok and @ro0NL
